### PR TITLE
Goal: Introduce new command for simulate

### PR DESCRIPTION
**Pull request by ahangsu**
## Summary

Introduce `goal clerk simulate` for simulation feature.

## Test Plan

e2e-sub test

### CODE REVIEW

**Comment by bbroder-algo** [comment]
seems like the right way forward

File: libgoal/libgoal.go [hunk 1]
```diff
@@ -1270,6 +1269,19 @@ func (c *Client) Dryrun(data []byte) (resp model.DryrunResponse, err error) {
...
 }
 
+// TransactionSimulation takes raw transaction or raw transaction group, and returns relevant simulation results.
+func (c *Client) TransactionSimulation(data []byte) (resp model.SimulateResponse, err error) {
+	algod, err := c.ensureAlgodClient()
+	if err != nil {
+		return
+	}
+	resp, err = algod.RawTransactionSimulate(data)
+	if err != nil {
```
Comment (barnjamin) [inline libgoal/libgoal.go:1279]: drop this?

Comment (ahangsu) [reply libgoal/libgoal.go:1279]: ouch, right

**Review by barnjamin — DISMISSED** [review]

File: test/scripts/e2e_subs/e2e-app-simulate.sh [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+${gcmd} clerk send -a 10000 -f ${ACCOUNT} -t ${ACCOUNT} -o pay1.tx
+${gcmd} clerk send -a 10000 -f ${ACCOUNT} -t ${ACCOUNT} -o pay2.tx
+
+cat pay1.tx pay2.tx | ${gcmd} clerk group -i - -o grouped.tx
+
+# We first test transaction group simulation WITHOUT signatures
+RES=$(${gcmd} clerk simulate -t grouped.tx | jq '."would-succeed"')
+
+if [[ $RES != $CONST_FALSE ]]; then
```
Comment (algochoi) [inline test/scripts/e2e_subs/e2e-app-simulate.sh:31]: Can we also check the `missing-signature` field here  too?

File: cmd/goal/clerk.go [hunk 3]
```diff
@@ -1253,6 +1257,34 @@ var dryrunRemoteCmd = &cobra.Command{
...
+	Use:   "simulate",
+	Short: "Simulate a transaction or transaction group offline",
+	Long:  `Simulate a transaction or transaction group offline under various conditions and verbosity.`,
+	Run: func(cmd *cobra.Command, args []string) {
+		data, err := readFile(txFilename)
+		if err != nil {
+			reportErrorf(fileReadError, txFilename, err)
+		}
+
+		dataDir := datadir.EnsureSingleDataDir()
```
Comment (algochoi) [inline cmd/goal/clerk.go:1270]: In dryrun, it looks like we also try to decode the txns and report a decoding error on the client side if this fails: https://github.com/algorand/go-algorand/blob/3e1772908f3643ceb818b3afca18da07e6f606bc/cmd/goal/clerk.go#L1126

Do we want to do something similar here, or just fall back on the server side error?

**Review by algochoi — COMMENTED** [review]
Looking good! Minor testing comment: might be interesting to also test something with a failure message (e.g. overspend).

File: test/scripts/e2e_subs/e2e-app-simulate.sh (hunk 2 above)
Comment (ahangsu) [reply test/scripts/e2e_subs/e2e-app-simulate.sh:31]: yes, adding a set of tests here

File: cmd/goal/clerk.go (hunk 3 above)
Comment (ahangsu) [reply cmd/goal/clerk.go:1270]: I think we send the transaction to server, then by following line
https://github.com/algorand/go-algorand/blob/master/daemon/algod/api/server/v2/handlers.go#L955

The decoding error should fail on server side.

File: test/scripts/e2e_subs/e2e-app-simulate.sh (hunk 2 above)
Comment (ahangsu) [reply test/scripts/e2e_subs/e2e-app-simulate.sh]: see [missing signature check](https://github.com/algorand/go-algorand/pull/5213/commits/d10ed71057a26579777a7b8af28c26ed92d266b0)

File: daemon/algod/api/client/restClient.go [hunk 4]
```diff
@@ -629,6 +630,12 @@ func (client RestClient) RawDryrun(data []byte) (response []byte, err error) {
 	return
 }
 
+// RawTransactionSimulate gets the raw transaction or raw transaction group, and returns relevant simulation results.
+func (client RestClient) RawTransactionSimulate(data []byte) (response model.SimulateResponse, err error) {
```
Comment (jasonpaulos) [inline daemon/algod/api/client/restClient.go:634]: nit: I'd suggest a name like `SimulateRawTransaction` instead. Functions with the `Raw` prefix in this file seem to return undecoded msgpack response bytes

**Review by jasonpaulos — COMMENTED** [review]
Generally looks good, and nice tests

File: daemon/algod/api/client/restClient.go (hunk 4 above)
Comment (ahangsu) [reply daemon/algod/api/client/restClient.go]: see https://github.com/algorand/go-algorand/pull/5213/commits/478d81629fdefab3996d94b81f09f96e5227e0b3

File: cmd/goal/clerk.go [hunk 5]
```diff
@@ -1253,6 +1257,34 @@ var dryrunRemoteCmd = &cobra.Command{
 	},
 }
 
+var simulateCmd = &cobra.Command{
+	Use:   "simulate",
+	Short: "Simulate a transaction or transaction group offline",
+	Long:  `Simulate a transaction or transaction group offline under various conditions and verbosity.`,
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply cmd/goal/clerk.go]: see https://github.com/algorand/go-algorand/pull/5213/commits/478d81629fdefab3996d94b81f09f96e5227e0b3

**Review by bbroder-algo — APPROVED** [review]

**Pull request merged**